### DISCUSSION TITLE
Status of MIDI and timemap expansion

### DISCUSSION THREAD

**Discussion by infojunkie** [rism-digital/verovio#3358] [Q&A] [answered]
I've been following the various issues regarding generation of MIDI and timemap with expanded repeats and jumps, but I am not sure what the current status is, nor what the roadmap is for those two features. [My own experiment](https://github.com/infojunkie/musicxml-player/blob/9ab87bd4311afe8173305d2df5b86b25f9ee86e6/src/VerovioConverter.ts#L19-L30) is not producing an expanded MIDI, but I am not sure I am doing it right.

Can anyone help clarify? Thanks in advance!

**Comment by lpugin** [comment]
It is not working because you are using MusicXML data. The time map expansion implementation relies on MEI xml:ids, meaning that it can work only if you load MEI data.

**Comment by infojunkie** [comment]
Thanks for your reply. As a stop-gap measure, I've implemented an [XSL transformation that expands a MusicXML score to an expanded one](https://github.com/infojunkie/musicxml-midi/blob/main/src/xsl/unroll.xsl). 

Is MusicXML support for expansion on your roadmap?

**Reply by lpugin** [reply]
Not really since this is essentially a MusicXML processing issue. It would be essentially a modification to the MusicXML importer for which a PR can be made.

**Comment by infojunkie** [comment]
I submitted PR https://github.com/rism-digital/verovio/pull/4250 to support MusicXML expansion.

**Comment by lpugin** [comment]
@infojunkie PR now has a `--expand-first` option, which I suggested to add and that processes the first `expansion` element in the score.

I am now thinking that we could do it differently. Since the most typical use case is to expand the MIDI and timemap, but not the SVG, we could have that set by default. That is, `--expand-first` not to be necessary for MIDI and timemap output. So basically I would replace it with two options:
* `--expand-enable-svg`, which would have to be set to render the expanded version of the score. It would expand the first `expansion`, unless another one is selected with `--expand id123`.
* `--expand-disable-midi`, which would disable any expansion in MIDI and timemap (basically the default we have now).

Any thoughts?

**Reply by samuelbradshaw** [reply]
Do you think it would be confusing to have one option be enable and the other option be disable? Would it make sense to do `--expand-enable-svg` (default false) and `--expand-enable-midi` (default true)?

If we do this, will an `<expansion>` element still be created for a score that doesn't have one? I've been looking into using the `plist` on the `<expansion>` element as a reference to manually create an expanded MEI score in JavaScript, then stripping out the coda symbols, special barlines, etc. in the expanded MEI score, then rendering to SVG, instead of using the `--expand` option, because the built-in `--expand` option doesn't clean up the coda symbols, barlines, etc.

**Reply by samuelbradshaw** [reply]
Maybe there could be an additional `--expand-enable-mei` option that generates the expanded MEI.

**Reply by samuelbradshaw** [reply]
Or maybe more simply, three boolean options: `--expand-svg`, `--expand-midi`, `--expand-mei`

**Reply by lpugin** [reply]
> Do you think it would be confusing to have one option be enable and the other option be disable? 

No, I think this makes it clearer what the default it.

> Would it make sense to do `--expand-enable-svg` (default false) and `--expand-enable-midi` (default true)?

That would not work because boolean options are always false (unset) by default

> Maybe there could be an additional `--expand-enable-mei` option that generates the expanded MEI.

That is not needed because `-t mei` would do it.

Another solution would be to have `--expand-enable` and `--expand-disable` and to explain that expansions are, by default, enabled with MIDI and timemap outputs, and disabled with SVG and MEI outputs. If a user does `--expand-disable -t mei`, or `--expand-enable -t midi`, they would just have no effect - which I think is fine.

**Reply by samuelbradshaw** [reply]
Nice, I wasn't aware of the -t / --output-to option! That will be useful. And good point about command line options being false if not set.

**Reply by samuelbradshaw** [reply]
It looks like -t / --output-to isn't supported in the JavaScript and Python toolkits, though. 🙁

**Reply by samuelbradshaw** [reply]
Sorry, I now see that if I call the toolkit method `getMEI()` _after_ using the `expand` option, the expanded MEI is returned, which is equivalent to the -t command line option.

**Reply by infojunkie** [reply]
I also think that having opposite default behaviours and flags depending on the output type is confusing UX-wise. Even more so that other output types exist and are not mentioned here (e.g. `-t mei`). Silently ignoring combinations of directives when they don't apply (e.g. `--expand-disable -t mei`) is also confusing for users.

Additionally, this UX issue is compounded by the introduction of `--expand-generate`. 

My own suggestion is the following:
- In the absence of any mention of expansion, never expand
- `--expand` without argument expands the first expansion
- `--expand id` expands the specified expansion
- `--expand-generate` generates an expansion AND expands it

**Reply by lpugin** [reply]
>  Silently ignoring combinations of directives when they don't apply (e.g. `--expand-disable -t mei`) is also confusing for users.

We can emit a warning if that is desirable.

Supporting both `--expand` and `--expand id` is confusing (and not really possible) since one is an option (boolean) and one a parameter.

In any case, I realised we need to deal with the document having to be reloaded for switching from expanded to un-expanded versions of the rendering, so I need to think about how that could work in the JS version of the toolkit.

**Comment by craigsapp** [comment]
As a negative option, `--expand-disable` is confusing (and particularly not backwards-compatible with behavior before `--expand` was implemented), making it both gratuitous and problematic. In addition, `--expand-first` has no clear meaning without a definition (is it a preferred ordering of the data? which is fragile and difficult to enforce in automatic processing, hence "buggy").

Traditional music notation has two basic states, and these two domains are not mixed; they also have opposite default `--expand` states:

(1) *score / printed order:* presented as the music is ordered on the page.  
(2) *performance order:* presented as the music is ordered in performance, particularly with respect to repeats and first/second endings.

In case (1), you rarely want to expand the score (it requires more paper and higher cost), while in case (2), such as MIDI conversion, you generally want an expanded view, where first/second endings are handled correctly rather than played straight through (the current state of MIDI export, which is very annoying.

I propose that when `-t mei` or `-t svg`  (or anything other than MIDI) is used, the `--expand` option should be turned off by default. For MIDI output, however, `--expand` should be turned on by default (with `--no-expand` or something similar to cancel).

You never want to hear the first repeat play through incorrectly, for example:  
https://verovio.humdrum.org/?file=joplin/countryclub.krn

<img width="374" height="204" alt="Screenshot 2026-01-19 at 12 52 01 AM" src="https://github.com/user-attachments/assets/4f86744b-5005-4a32-a754-06207928a442" />

Currently, MIDI is exported in paper ordering, which is very annoying. It would be preferable for SVG and MIDI output to transparently handle this situation gracefully without requiring the user to explicitly set `--expand` or `--no-expand`.

Thus, when `-t mei` is used, the expand feature should be turned off by default. It is not practical to expand a paper score into performance order; for example, first/second endings need to be removed in the expanded output. This process is more about transforming an edition (to make another edition). I do this to create performance scores in the Josquin Research Project when page turns would otherwise be disrupted and therefore avoided.

---

This concept is related to the `thru` command in the Humdrum Toolkit.

Humdrum User’s Guide: *Electronic Music Editing: Editing Sections*  
https://www.humdrum.org/guide/ch37/#editing-sections

Chapter 20: *Strophes, Verses and Repeats*  
https://www.humdrum.org/guide/ch20

Chapter 37:  
https://www.humdrum.org/guide/ch37/#editing-sections

Reference documentation:  
https://www.humdrum.org/Humdrum/commands/thru.html

---

For an analytical application: Download example data and running `notecount` analyses:

<img width="810" height="490" alt="Screenshot 2026-01-19 at 12 41 35 AM" src="https://github.com/user-attachments/assets/03147d90-81ac-49e5-a228-c30a5a14e0cc" />

Download:
```bash
git clone https://github.com/craigsapp/mozart-piano-sonatas
cd mozart-piano-sonatas/kern
```

Note that there are always two performance orders:

(A) *full repeats*: performed as the repeat is written  
(B) *second endings*: skip repeats, playing only once for minimal performance

First, count the number of notes (excluding ties):

```bash
cat *.krn | notecount -s
```

Result: 96,635 notes, 97,613 noteheads (including tied notes), and 9,772 rests.

Performance as written:

```bash
cat *.krn | thrux | notecount -s
```

Result: 149,737 note attacks.

Avoiding repeats:

```bash
cat *.krn | thrux -v norep | notecount -s
```

Result: 97,202 note attacks, meaning that  
`(1 − 97,202 / 149,737) × 100 ≈ 35%` of the notes are repeated.

---

In MEI, this is represented by `<expansion>` elements:

```xml
<choice xml:id="v11l01xa">
  <orig xml:id="wlzhmjr">
    <expansion plist="#label-A #label-A #label-B"
               xml:id="expansion-L18F1"/>
  </orig>
  <reg xml:id="yc56qmy" type="norep">
    <expansion plist="#label-A #label-B"
               xml:id="expansion-L19F1"
               type="norep"/>
  </reg>
</choice>
```

Humdrum equivalents:

```tsv
*>[A,A,B]       *>[A,A,B]       *>[A,A,B]
*>norep[A,B]    *>norep[A,B]    *>norep[A,B]
```

"norep" means "no repeats" or a short form of "no: repeats: second (last) ending only".
Note that `[A,A,B]` is the "default" (unlabeled) expanion lists.  It does not necessarily need to be "first":

```tsv
*>norep[A,B]    *>norep[A,B]    *>norep[A,B]
*>[A,A,B]       *>[A,A,B]       *>[A,A,B]

```


Section labels:

```tsv
*>A   *>A   *>A
*>B   *>B   *>B
```

The default expansion list `A,A,B` copies section A (lines 20–664) twice before section B.

More complex cases (e.g., da capo minuets) are categorical and not well suited to ID-based reference. Using `expansion@label` instead avoids “magic” IDs and gives `--expand-first` a defined semantic meaning (unlabeled or empty "" name).

**Reply by samuelbradshaw** [reply]
To clarify, are you recommending that Verovio doesn't allow a user to export MIDI in paper order, and doesn't allow exporting MEI in performance order? Or just that those are defaults that can be overridden?

In a digital context, rendering the MEI in expanded/unrolled performance order is useful, especially for congregational singing where singers aren't trained musicians and don't necessarily know how to follow repeats, codas, etc. I am working on providing this as an option in my application (with post-processing to clean up repeat marks, barlines, lyrics, etc.).

My application also relies on the unexpanded MIDI currently. I may be able to refactor to use expanded MIDI, but the unexpanded MIDI is nice because it's a smaller file (smaller file transfer / download) that can be expanded on the fly after loading.

I don't feel strongly what the default should be, as long as those options remain available.

**Comment by craigsapp** [comment]
> Supporting both --expand and --expand id is confusing (and not really possible) since one is an option (boolean) and one a parameter.

`--expand ""` &rarr; `--expand id`, where the ID is empty (using `expansion@label` instead of `expansion@xml:id`, where "" is the "default" expansion.

**Comment by lpugin** [comment]
> In addition, `--expand-first` has no clear meaning without a definition (is it a preferred ordering of the data?)

Yes, exactly. The first you would get with and xpath looking for `./expansion`. I don't think this is particularly problematic, and you can always use `--expand xmlId` if you want.

later

> For MIDI output, however, `--expand` should be turned on by default (with `--no-expand` or something similar to cancel).

That's the same negative option as `--expand-disable`, so I am not sure I understand your position on this. It does seem that you agree that `-t svg-` and `-t mei` should produce un-expanded output by default, while `-t midi` and `-t timemap` expanded output. Am I right? If yes, I still think that having two options `--expand-enable` and `--expand-disable` is the clearest solution to have a way to change default behavior. If something like `--expand-always` and `--expand-never` is less confusing, I am fine with that too. I don't want to go with `--expand ""` since this will be problematic.

**Comment by infojunkie** [comment, marked as answer]
As per https://github.com/rism-digital/verovio/pull/4267 

> (default) will expand MIDI / timemap (but not SVG / MEI) by selecting the first expansion or trying to generate one
> --expand xmlid expand MIDI / timemap (but not SVG / MEI) will be the same as default but with a specific expansion selected (less > backward compatible but clearer)
> --expand-always will expand everything by selecting the first expansion or trying to generate one
> --expand xmlid --expand-always will expand everything with a specific expansion selected

Also, as per https://github.com/rism-digital/verovio/pull/4250, a MusicXML import will automatically generate an `expansion` element based on the repeats and jumps contained in the score.
